lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs

```diff
@@ -153,12 +193,290 @@ sqliteSpec = withDB (fst <$> newMemoryDBLayer) $ do
 
 sqliteSpecRnd :: Spec
 sqliteSpecRnd =
-    withDB (fst <$> newMemoryDBLayer @(RndState DummyTarget) @DummyTarget @RndKey) $ do
+    withDB newDBRnd $ do
         describe "Sqlite simple (RndState)" $
             simpleSpec testCpRnd
         describe "Sqlite State machine (RndState)" $ do
             it "Sequential state machine tests" prop_sequential
 
+withDBCtx
+    :: IO (SqliteContext, DBLayer IO s t k)
+    -> SpecWith (SqliteContext, DBLayer IO s t k)
+    -> Spec
+withDBCtx create = beforeAll create . beforeWith
+    (\(ctx, db) -> cleanDB db $> (ctx, db))
+
+multipleCheckpointsSpec
+    :: forall s k.
+       ( Show (k 'RootK XPrv)
+       , Eq (k 'RootK XPrv)
+       , Eq s
+       , IsOurs s
+       , NFData s
+       , Show s
+       , GenerateTestKey k )
+    => Wallet s DummyTarget
+    -> SpecWith (SqliteContext, DBLayer IO s DummyTarget k)
+multipleCheckpointsSpec initialState = do
+    describe "put and read checkpoints" $ do
+        it "previous checkpoints are also stored" $ \(ctx,db) -> do
```

Comment:
Hmmm... That's quite complicated :thinking: 
It'd be nice to outline what we're trying to "prove" here. If I understand correctly:

- We want to verify that we can store new checkpoints and preserve past ones. 
- I don't understand what others scenarios are trying to achieve :cry: (sorry)

To me, it seems like we have gone through a wrong path here and we need to step back a little. Also, I don't understand why we query the database directly instead of using the db layer. We build the db layer to construct also an invariant-preserving interface. We shouldn't have to use anything else; that's a path which could be quite error-prone. So let's stick with the DBLayer functions. We aren't testing against an implementation, we are testing against an interface!

Also, it seems that all tests contains a lot of arbitrary data ? Why not generating transactions and checkpoints (we already have code for that in the `DBSpec` module!).  We could come up with properties in the very same way we've done it before. For example: calling putCheckpoint `n` times, and `readcheckpoint` should give us the last entry; actually we already have such property :smile: (cf `prop_sequentialPut` !) 

I think we have to start over with testing here and, begin with stating what we want to test. Then we can think about what's the most suitable way of testing it (unit, property, state-machine, integration ...)